release-23.2: sql/schemachanger: Fix early exit for IF NOT EXISTS when adding existing columns

spilchen opened the pull request:
Backport 1/1 commits from #137633.

/cc @cockroachdb/release

---

Previously, we exited too early when adding an existing column with the IF NOT EXISTS option. This skipped necessary AST handling, such as annotating resolved names, which subsequent processing depended on. To fix this, the existing column check has been moved to just before adding or removing SCPB elements.

Additionally, we failed to handle cases where the column had been dropped in a prior statement of an explicit transaction. Previously, the operation would succeed without adding any elements. Now, the column will be correctly added.

Epic: None
Closes #137326
Release note (bug fix): Fixed an issue where adding an existing column with the IF NOT EXISTS option could exit too early, skipping necessary handling of the AST. This could lead to statement failure of the ALTER.
Release justification: low risk fix that improves ADD COLUMN

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137678)
<!-- Reviewable:end -->

spilchen committed:
sql/schemachanger: Fix early exit for IF NOT EXISTS when adding existing columns

Previously, we exited too early when adding an existing column with the
IF NOT EXISTS option. This skipped necessary AST handling, such as
annotating resolved names, which subsequent processing depended on. To
fix this, the existing column check has been moved to just before adding
or removing SCPB elements.

Additionally, we failed to handle cases where the column had been
dropped in a prior statement of an explicit transaction. Previously, the
operation would succeed without adding any elements. Now, the column will be
correctly added.

Epic: None
Closes #137326
Release note (bug fix): Fixed an issue where adding an existing column
with the IF NOT EXISTS option could exit too early, skipping necessary
handling of the AST. This could lead to statement failure of the ALTER.
